### PR TITLE
Break archive operation into batches

### PR DESCRIPTION
**Pull request by WilliamBZA**
Connects to https://github.com/Particular/PlatformDevelopment/issues/1164

Renaming all Retry specific code to be generic operations so as to support both archive as well as retries.

Ping @WojcikMike

### CODE REVIEW

File: src/ServiceControl/Infrastructure/RavenDB/RavenBootstrapper.cs [hunk 1]
```diff
@@ -48,7 +48,7 @@ public void StartRaven(EmbeddableDocumentStore documentStore, Settings settings,
             Directory.CreateDirectory(settings.DbPath);
 
             documentStore.DataDirectory = settings.DbPath;
-            documentStore.UseEmbeddedHttpServer = maintenanceMode || settings.ExposeRavenDB;
+            documentStore.UseEmbeddedHttpServer = true;// maintenanceMode || settings.ExposeRavenDB;
```
Comment (WojcikMike) [inline src/ServiceControl/Infrastructure/RavenDB/RavenBootstrapper.cs]: Why this have changed? was it intentional?

Comment (WilliamBZA) [inline src/ServiceControl/Infrastructure/RavenDB/RavenBootstrapper.cs]: Nope, that's so I can query raven while I'm working

**Comment by gbiellem** [comment]
@WilliamBZA  In the registry under `HKEY_LOCAL_MACHINE\SOFTWARE\ParticularSoftware\ServiceControl` add a string `ExposeRavenDB` with a value of `true` and you'll always have HTTP enabled with a code change

File: src/ServiceControl/Infrastructure/RavenDB/RavenBootstrapper.cs (hunk 1 above)
Comment (WojcikMike) [inline src/ServiceControl/Infrastructure/RavenDB/RavenBootstrapper.cs]: I would suggest so that we revert this change and use @gbiellem [solution](https://github.com/Particular/ServiceControl/pull/934#issuecomment-286667354)

**Comment by WilliamBZA** [comment]
Ping @Particular/servicecontrol-maintainers and @janovesk, this is ready for review now.

**Comment by WilliamBZA** [comment]
Ping @Particular/servicecontrol-maintainers for a review again.

If there are no comments by Wednesday morning I will merge and release.

**Comment by WilliamBZA** [comment]
And ping @janovesk again

File: src/ServiceControl/Recoverability/Grouping/Handlers/ArchiveAllInGroupHandler.cs [hunk 2]
```diff
@@ -1,83 +1,101 @@
...
+                    }
+                    
+                    logger.Info($"Splitting group {message.GroupId} into batches");
+                    archiveOperation = documentManager.CreateArchiveOperation(session, message.GroupId, ArchiveType.FailureGroup, message.CutOff, groupDetails.NumberOfMessagesInGroup, groupDetails.GroupName, batchSize);
+                    session.SaveChanges();
+                }
             }
 
-            var groupName = "Undefined";
+            archiveOperationManager.StartArchiving(archiveOperation.RequestId, archiveOperation.ArchiveType, archiveOperation.TotalNumberOfMessages, archiveOperation.NumberOfMessagesArchived, archiveOperation.Started, archiveOperation.GroupName, archiveOperation.NumberOfBatches, archiveOperation.CurrentBatch);
```
Comment (janovesk) [inline src/ServiceControl/Recoverability/Grouping/Handlers/ArchiveAllInGroupHandler.cs]: Probably just pass in the archiveOperation object here instead?

File: src/ServiceControl/Recoverability/Retries/RetriesGateway.cs [hunk 3]
```diff
@@ -20,7 +20,7 @@ public class RetriesGateway
 
         private IDocumentStore store;
         private RetryDocumentManager retryDocumentManager;
-        public RetryOperationManager RetryOperationManager { get; set; }
+        public OperationManager RetryOperationManager { get; set; }
```
Comment (janovesk) [inline src/ServiceControl/Recoverability/Retries/RetriesGateway.cs]: Rename to `OperationManager` now that it is generic.

File: src/ServiceControl/Recoverability/Retries/RetryDocumentManager.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@ namespace ServiceControl.Recoverability
     public class RetryDocumentManager
     {
         protected static string RetrySessionId = Guid.NewGuid().ToString();
-        public RetryOperationManager RetryOperationManager { get; set; }
+        public OperationManager RetryOperationManager { get; set; }
```
Comment (janovesk) [inline src/ServiceControl/Recoverability/Retries/RetryDocumentManager.cs]: Rename to `OperationManager` now that it is generic.

File: src/ServiceControl/Recoverability/Retries/RetryProcessor.cs [hunk 5]
```diff
@@ -29,7 +29,7 @@ class RetryProcessor
 
         static ILog Log = LogManager.GetLogger(typeof(RetryProcessor));
 
-        public RetryProcessor(ISendMessages sender, IBus bus, ReturnToSenderDequeuer returnToSender, RetryOperationManager retryOperationManager)
+        public RetryProcessor(ISendMessages sender, IBus bus, ReturnToSenderDequeuer returnToSender, OperationManager retryOperationManager)
```
Comment (janovesk) [inline src/ServiceControl/Recoverability/Retries/RetryProcessor.cs]: Rename to `OperationManager` now that it is generic.

File: src/ServiceControl/Recoverability/Retries/RetryProcessor.cs [hunk 6]
```diff
@@ -218,7 +218,7 @@ void StageMessage(FailedMessage message, string stagingId)
         ISendMessages sender;
         IBus bus;
         ReturnToSenderDequeuer returnToSender;
-        RetryOperationManager retryOperationManager;
+        OperationManager retryOperationManager;
```
Comment (janovesk) [inline src/ServiceControl/Recoverability/Retries/RetryProcessor.cs]: Rename to `OperationManager` now that it is generic.

File: src/ServiceControl/Recoverability/Retries/UnacknowledgedRetryOperation.cs [hunk 7]
```diff
@@ -5,7 +5,7 @@
     public class UnacknowledgedRetryOperation
     {
         public string RequestId { get; set; }
-        public RetryType RetryType { get; set; }
+        public RetryType OperationType { get; set; }
```
Comment (janovesk) [inline src/ServiceControl/Recoverability/Retries/UnacknowledgedRetryOperation.cs]: Seems off. OpertationType of type RetryType?

File: src/ServiceControl/Recoverability/Archiving/ArchiveOperationLogic.cs [hunk 8]
```diff
@@ -0,0 +1,128 @@
+﻿namespace ServiceControl.Recoverability
+{
+    using ServiceControl.Infrastructure.DomainEvents;
+    using System;
+
+    public class ArchiveOperationLogic
```
Comment (janovesk) [inline src/ServiceControl/Recoverability/Archiving/ArchiveOperationLogic.cs]: Isn't this just a ArchiveOperation? "Logic" is one of those fuzzy words that don't add anything concrete to me.

File: src/ServiceControl/Recoverability/Grouping/DomainHandlers/StoreHistoryHandler.cs [hunk 9]
```diff
@@ -35,7 +35,7 @@ public void Handle(RetryOperationCompleted message)
                 retryHistory.AddToUnacknowledged(new UnacknowledgedRetryOperation
                 {
                     RequestId = message.RequestId,
-                    RetryType = message.RetryType,
+                    OperationType = message.RetryType,
```
Comment (WilliamBZA) [inline src/ServiceControl/Recoverability/Grouping/DomainHandlers/StoreHistoryHandler.cs]: To distinguish the retry type. Group, single message, multiple messages, etc. I don't think it's actually used anywhere though.

Comment (janovesk) [inline src/ServiceControl/Recoverability/Grouping/DomainHandlers/StoreHistoryHandler.cs]: If Operation is a generalisation of Retry and Archive, I would expect something called OperationType to consists of either `Retry` or `Archive`, not `Single Message`, `Group`, `All`  or `Multiple messages`.

Comment (WilliamBZA) [inline src/ServiceControl/Recoverability/Grouping/DomainHandlers/StoreHistoryHandler.cs]: Hmm, we can't actually change this class at all or the previously stored documents won't be loadable.

File: src/ServiceControl/Recoverability/Archiving/ArchiveOperationLogic.cs (hunk 8 above)
Comment (WilliamBZA) [inline src/ServiceControl/Recoverability/Archiving/ArchiveOperationLogic.cs]: The problem is we store the `ArchiveOperation` as the header operation - which has the same name as this should have been. I'm open for other suggestions though?
